api/lightning/node_cln.py

```diff
@@ -0,0 +1,408 @@
+            "remote_balance": remote_balance_sat,
+            "unsettled_local_balance": 0,
+            "unsettled_remote_balance": 0,
+        }
+
+    @classmethod
+    def pay_onchain(cls, onchainpayment, queue_code=5, on_mempool_code=2):
+        """Send onchain transaction for buyer payouts"""
+
+        if DISABLE_ONCHAIN or onchainpayment.sent_satoshis > MAX_SWAP_AMOUNT:
+            return False
+
+        request = noderpc.WithdrawRequest(
+            destination=onchainpayment.address,satoshi=int(onchainpayment.sent_satoshis),
+            feerate=str(int(onchainpayment.mining_fee_rate)*1_000)+"perkb",
+            minconf=int(not config("SPEND_UNCONFIRMED", default=False, cast=bool))
+        )
+        
+        # Cheap security measure to ensure there has been some non-deterministic time between request and DB check
+        delay = (
+            secrets.randbelow(2**256) / (2**256) * 10
+        )  # Random uniform 0 to 5 secs with good entropy
+        time.sleep(3 + delay)
+
+        if onchainpayment.status == queue_code:
+            # Changing the state to "MEMPO" should be atomic with SendCoins.
+            onchainpayment.status = on_mempool_code
+            onchainpayment.save()
+            response = cls.stub.Withdraw(request)
+
+            if response.txid:
+                onchainpayment.txid = response.txid
+                onchainpayment.broadcasted = True
+            onchainpayment.save()
+            return True
+
+        elif onchainpayment.status == on_mempool_code:
+            # Bug, double payment attempted
+            return True
+
+    @classmethod
+    def cancel_return_hold_invoice(cls, payment_hash):
+        """Cancels or returns a hold invoice"""
+        request = noderpc.HodlInvoiceCancelRequest(payment_hash=bytes.fromhex(payment_hash))
+        response = cls.stub.HodlInvoiceCancel(request)
+        
+        return response.state == 1  # True if state is CANCELED, false otherwise.
+
+    @classmethod
+    def settle_hold_invoice(cls, preimage):
+        """settles a hold invoice"""
+        request = noderpc.HodlInvoiceSettleRequest(payment_hash=hashlib.sha256(bytes.fromhex(preimage)).digest())
+        response = cls.stub.HodlInvoiceSettle(request)
+        
+        return response.state == 2  # True if state is SETTLED, false otherwise.
+
+    @classmethod
+    def gen_hold_invoice(
+        cls, num_satoshis, description, invoice_expiry, cltv_expiry_blocks
+    ):
+        """Generates hold invoice"""
+
+        hold_payment = {}
+        # The preimage is a random hash of 256 bits entropy
+        preimage = hashlib.sha256(secrets.token_bytes(nbytes=32)).digest()
+
+        # Its hash is used to generate the hold invoice
+        r_hash = hashlib.sha256(preimage).digest()
+
+        request = noderpc.InvoiceRequest(
+            description=description,
+            amount_msat=num_satoshis * 1_000,
+            label="TODO", # TODO needs to be a unique string
```

Comment:
A "label" in CLN is some sort of description only for internal use? We could add a couple more parameters to `gen_hodl_invoice()` so this function is informed on `order_id` , `receiver_robot`, `time` and other unique identifiers to build a string for this label that is useful for bookeeping too.

Reply:
done